### PR TITLE
Enhance metadata of sharepointLoader.

### PR DESCRIPTION
**Pull request by rahul-trip**
Description: 2 feature flags added to SharePointLoader in this PR:

1. load_auth: if set to True, adds authorised identities to metadata
2. load_extended_metadata, adds source, owner and full_path to metadata

Unit tests:N/A
Documentation: To be done.

### CODE REVIEW

**Comment by isahers1** [comment]
@rahul-trip Ping us when ready for review!

**Comment by rahul-trip** [comment]
Hi @isahers1,

Kindly review these changes. Thanks!

**Comment by isahers1** [comment]
> Hi @isahers1,
> 
> Kindly review these changes. Thanks!

Will try to take a look tomorrow!

File: libs/community/langchain_community/document_loaders/sharepoint.py [hunk 1]
```diff
@@ -90,38 +117,58 @@ def lazy_load(self) -> Iterator[Document]:
...
+                        group_names.append(email)
         return group_names
 
     def _fetch_access_token(self) -> Any:
         with open(self.token_path) as f:
             s = f.read()
         data = json.loads(s)
         return data
+
+    def get_extended_metadata(self, file_id: str) -> dict:
```
Comment (isahers1) [inline libs/community/langchain_community/document_loaders/sharepoint.py:152]: I am new to Sharepoint, so I don't know too much about it's usability - why is this function helpful/will be used by users?

File: libs/community/langchain_community/document_loaders/sharepoint.py [hunk 2]
```diff
@@ -90,38 +117,58 @@ def lazy_load(self) -> Iterator[Document]:
...
+        data = self._fetch_access_token()
+        access_token = data.get("access_token")
```
Comment (isahers1) [inline libs/community/langchain_community/document_loaders/sharepoint.py:154]: Is data guaranteed to have an "access_token" field? I could see the self.token_path file containing the wrong key maybe

File: libs/community/langchain_community/document_loaders/sharepoint.py [hunk 3]
```diff
@@ -37,6 +37,8 @@ class SharePointLoader(O365BaseLoader, BaseLoader):
     """ The ID of the file for which we need auth identities"""
     site_id: Optional[str] = None
     """ The ID of the Sharepoint site of the user where the file is present """
+    load_extended_metadata: Optional[bool] = False
```
Comment (isahers1) [inline libs/community/langchain_community/document_loaders/sharepoint.py:40]: New to Sharepoint, so might be a basic question: but why would people want this additional metadata?

File: libs/community/langchain_community/document_loaders/sharepoint.py (hunk 1 above)
Comment (isahers1) [inline libs/community/langchain_community/document_loaders/sharepoint.py:152]: Can you add a doctoring?

**Review by isahers1 — COMMENTED** [review]
Left a few comments, will check back before end of week

File: libs/community/langchain_community/document_loaders/sharepoint.py (hunk 2 above)
Comment (rahul-trip) [reply libs/community/langchain_community/document_loaders/sharepoint.py:154]: Hi @isahers1 
Yes it is always expected to have `access_token` field (On successful auth) as the authentication happens before data loading. And it is expected that the 2 step authentication would store this file at ~/.credentials/o365_token.txt

please refer para 2nd and 3rd for more info:
https://python.langchain.com/v0.2/docs/integrations/document_loaders/microsoft_sharepoint/#-authentication

File: libs/community/langchain_community/document_loaders/sharepoint.py (hunk 3 above)
Comment (rahul-trip) [reply libs/community/langchain_community/document_loaders/sharepoint.py:40]: It is an enhancement to the loader. As of today SharepointLoader is not very rich in metadata.
I have added 3 of the fields which makes the Documents metadata ready for consumption for any app.

1. size: size of the source file
2. owner: owner of the source file
3. full_path: pretty human readable path of the source file.

Note: Please refer GoogleDriveLoader with similar signatures: https://github.com/langchain-ai/langchain-google/blob/6495426d7db176fa72803abf9a356d5083bf19c5/libs/community/langchain_google_community/drive.py#L50

File: libs/community/langchain_community/document_loaders/sharepoint.py (hunk 1 above)
Comment (rahul-trip) [reply libs/community/langchain_community/document_loaders/sharepoint.py:175]: Added and updated Docstrings throughout the file.

**Pull request merged**